File: QytetetJava/src/modeloqytetet/Jugador.java
package modeloqytetet;

import java.util.ArrayList;
        
        
public class Jugador implements Comparable {
   private Boolean encarcelado;
   private String nombre;
   private int saldo;
   private Casilla casillaActual;
   private Sorpresa cartaLibertad;
   private ArrayList<TituloPropiedad> propiedades;
   
  
   public Jugador(String nombre){
       encarcelado=false;
       this.nombre=nombre;
       saldo=7500;
       this.cartaLibertad = null;
       this.casillaActual = null;
       this.propiedades = new ArrayList<TituloPropiedad>();
   }
   
    @Override
    public int compareTo(Object otroJugador) {
        int otroCapital = ((Jugador) otroJugador).obtenerCapital();
        return otroCapital - obtenerCapital();
    }

   @Override
     public String toString(){
         String propiedadesJugador ="";
         for(int i=0;i<this.propiedades.size();i++){
                               propiedadesJugador+= this.propiedades.get(i).toString()+"\n";
         }
         if (this.propiedades.size()==0){
            return "Nombre: "+this.nombre + "\n"+
                   "Saldo: "+this.saldo + "\n"+
                   "Capital: "+obtenerCapital() +"\n"+
                   "Encarcelado: "+ this.encarcelado + "\n"+
                   "CasillaActual: "+this.casillaActual +"\n"+
                   "CartaLibertad: "+this.cartaLibertad +"\n";
                    
         }else
             return "Nombre: "+this.nombre + "\n"+
                    "Saldo: "+this.saldo + "\n"+
                    "Capital: "+obtenerCapital() +"\n"+
                    "Encarcelado: "+ this.encarcelado + "\n"+
                    "CasillaActual: "+this.casillaActual +"\n"+
                    "CartaLibertad: "+this.cartaLibertad +"\n"+
                    "Propiedades: "+ propiedadesJugador + "\n";
                 
                
     }
    
   //Boolean cancelarHipoteca(TituloPropiedad titulo){}
   
   //Boolean comprarTitulPropiedad(){}
   
   int cuantasCasasHotelesTengo(){
       int sumaHotelesCasas=0;
       for(int i=0; i< this.propiedades.size();i++){
           sumaHotelesCasas += this.propiedades.get(i).getNumCasas();
           sumaHotelesCasas +=this.propiedades.get(i).getNumHoteles();
        }
        return sumaHotelesCasas;
   }
   
   //Boolean deboPagarAlquiler(){}
   
   Sorpresa devolverCartaLibertad(){
       Sorpresa libertadTemporal;
       libertadTemporal=new Sorpresa(cartaLibertad.getTexto(),
               cartaLibertad.getValor(),cartaLibertad.getTipoSorpresa());
       
       cartaLibertad=null;
       
       return libertadTemporal;
   }
   
   Boolean edificarCasa(TituloPropiedad titulo){
        int numcasas=titulo.getNumCasas();
        Boolean devolver=false;
        if(titulo.getPropietario().esDeMiPropiedad(titulo)==true){
            if(numcasas<4){
                titulo.setCasas(numcasas+1);
                devolver=true;
            }
        }
        return devolver;
   }
   
   Boolean edificarHotel(TituloPropiedad titulo){
       int numcasas=titulo.getNumCasas();
       int numhoteles=titulo.getNumHoteles();
       Boolean devolver=false;
       
       if(titulo.getPropietario().esDeMiPropiedad(titulo)==true){
           if(numcasas==4){
               titulo.setCasas(0);
               titulo.setHoteles(1);
               devolver=true;
           }
       }
       return devolver;
   }
   
   //private void eliminarDeMisPropiedades(){}
   
   protected Boolean esDeMiPropiedad(TituloPropiedad titulo){
       Boolean encontrado=false;
       
       for(int i=0;i<this.propiedades.size() && encontrado==false ;i++){
           if(this.propiedades.get(i).getNombre() == titulo.getNombre()){
               encontrado=true;
           }
       }
       return encontrado;
   }
   
   Boolean estoyEnCalleLibre(){
        Boolean devolver=true;
        
        if(casillaActual.getTituloPropiedad().tengoPropietario()==true)
            devolver=false;
        
        return devolver;
   }
   
   Sorpresa getCartaLibertad(){
        return cartaLibertad;
   }
   
   Casilla getCasillaActual(){
       return this.casillaActual;
   }
   
   Boolean getEncarcelado(){
       return encarcelado;
   }
   
   public String getNombre(){
       return nombre;
   }
   
   public ArrayList<TituloPropiedad> getPropiedades(){
       return propiedades;
   }
   
   public int getSaldo(){
       return saldo;
   }
   
   Boolean hipotecarPropiedad(TituloPropiedad titulo){
       Boolean hipotecado=false;
       if(this.esDeMiPropiedad(titulo)==true){
           if(titulo.getHipotecada()==false){
               titulo.cambiaHipotecada();
               hipotecado=true;
            }
       }
       return hipotecado;
   }
   
   //void irACarcel(Casilla casilla){}
   
   int modificarSaldo(int cantidad){
       this.saldo-=cantidad;
       return saldo;
}
   
   int obtenerCapital(){
        int capital = this.saldo;
            for(int  i = 0; i < this.propiedades.size(); i++){
                int precioEdificarHotelyCasa = this.propiedades.get(i).getPrecioEdificar();
                int costeBase = this.propiedades.get(i).getPrecioCompra();
                int numCasas = this.propiedades.get(i).getNumCasas();
                int numHoteles = this.propiedades.get(i).getNumHoteles();
  
                if(!this.propiedades.get(i).getHipotecada()){
                    capital += costeBase * (numCasas + numHoteles);
                }else{
                    capital -= costeBase * (numCasas + numHoteles);
            }
        }
        
        return capital;
   }
   
   ArrayList<TituloPropiedad> obtenerPropiedades(Boolean estadoHipoteca){
       
       ArrayList<TituloPropiedad> hipotecadas = new ArrayList<TituloPropiedad>();
       ArrayList<TituloPropiedad> noHipotecadas = new ArrayList<TituloPropiedad>(); 
        
       for(int i = 0; i < this.propiedades.size(); i++){
            if(this.propiedades.get(i).getHipotecada() == true)
                hipotecadas.add(this.propiedades.get(i));
            else
                noHipotecadas.add(this.propiedades.get(i));
}
       if(estadoHipoteca == true)
           return hipotecadas;
       else
           return noHipotecadas;
       
   }
   
   //void pagarAlquiler(){}
   
   void pagarImpuesto(){
       saldo -= casillaActual.getCoste();
    }
   
   void pagarLibertad(int  cantidad){
       if(encarcelado == true)
           if(this.saldo>=cantidad);
               this.setSaldo(saldo-cantidad);
   }
   
   //void setAttribute(attribute){}
   
   void setCartaLibertad(Sorpresa carta){
       this.cartaLibertad=carta;
   }
   
   void setCasillaActual(Casilla casilla){
       this.casillaActual = casilla;
   }
   
   void setEncarcelado(Boolean encarcelado){
       this.encarcelado=encarcelado;
   }

   
   void setSaldo(int saldo){
       this.saldo=saldo;
   }
   
   Boolean tengoCartaLibertad(){
       if(cartaLibertad != null){
           return true;
       }else
           return false;
   }
   
   private Boolean tengoSaldo(int cantidad){
       if(saldo >= cantidad)
           return true;
       else 
           return false;
   }
   
   Boolean venderPropiedad(Casilla casilla){
       int numerocasilla=casilla.getNumeroCasilla();
       Boolean devolver=false;
       if(this.esDeMiPropiedad(this.propiedades.get(numerocasilla))==true){
           this.getPropiedades().remove(numerocasilla);
           this.setSaldo(this.getPropiedades().get(numerocasilla).getPrecioCompra());
           devolver=true;
       }  
       return devolver;    
   }
   
   
}

File: Qytetet_Ruby/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

class Jugador
  attr_accessor:encarcelado
  attr_accessor:nombre
  attr_accessor:saldo
  attr_accessor:carta_libertad
  attr_accessor:propiedades
  attr_accessor:casilla_actual
  
  def initialize(unNombre)
    @encarcelado=false
    @nombre=unNombre
    @saldo=7500
    @propiedades = Array.new
  end
  
  #def cancelar_hipoteca
  
  #end
  
  #def comprar_titulo_propiedad
  
  #end
  
  def cuantas_casas_hoteles_tengo
    devolver = 0
    for i in 0..@propiedades.size()
      devolver += @propiedades[i].get_numcasas()
      devolver += @propiedades[i].get_numhoteles()
    end
    return devolver
  end
  
  #def debo_pagar_alquiler
  
  #end
  
  def devolver_carta_libertad
    devolver = Sorpresa.new(@carta_libertad.get_texto,@carta_libertad.get_tipo,@carta_libertad.get_valor)
    @carta_libertad=null
    return devolver
  end
  
  #def edificar_casa(unTituloPropiedad)
  
  #end
  
  #def edificar_hotel(unTituloPropiedad)
  
  #end
  
  #def eliminar_de_mis_propiedades(unTituloPropiedad)
  
  #end
  
  def es_de_mi_propiedad(unTituloPropiedad)
    devolver = false
    for i in 0..@propiedades.size
      if(@propiedades[i].get_nombre()==unTituloPropiedad.get_nombre())
        devolver = true
      end
    end
    return devolver
  end
  
  #def estoy_en_calle_libre
  
  #end
  
  def get_carta_libertad
   return @carta_libertad
  end
  
  def get_casilla_actual
    return @casilla_actual
  end
  
  def get_encarcelado
    return @encarcelado
  end
  
  def get_nombre
    return @nombre
  end
  
  def get_propiedades
    return @propiedades
  end
  
  def get_saldo
    return @saldo
  end
  
  #def hipotecar_propiedad(unTituloPropiedad)
  
  #end
  
  #def ir_a_carcel(Casilla)
  
  #end
  
  def modificar_saldo(cantidad)
    @saldo += cantidad
    return @saldo
  end
  
  def obtener_capital
    devolver = @saldo
    for i in 0..@propiedades.size()
      if(@propiedades[i].get_hipotecada)
        devolver -= @propiedades[i].get_hipoteca_base
      else
        devolver += @propiedades[i].get_precio_compra()
        devolver += @propiedades[i].get_numcasas() * @getpropiedades[i].get_factor_revalorizacion()
      end
    end
    return devolver
  end
  
  def obtener_propiedades(hipotecada)
    devolver = Array.new
    for i in 0..@propieades.size()
      if(@propieades[i].get_hipotecada==hipotecada)
        devolver<<@propiedades[i]
      end
    end
    return devolver
  end
  
  #def pagar_alquiler
  
  #end
  
  def pagar_impuesto
    @saldo -= @casilla_actual.get_coste
  end
  
  #def pagar_libertad(cantidad)
  
  #end
  
  def set_carta_libertad(carta)
    @carta_libertad=Sorpresa.new("Los guardias de las mazmorras se han quedado"\
              " dormidos y consigues escapar,quedas libre de las mazmorras.",TipoSorpresa::SALIRCARCEL,0)
  end
  
  def set_casilla_actual(casilla)
    @casilla_actual=casilla
  end
  
  def set_encarcelado(encarcelado=false)
    @encarcelado=encarcelado
  end
  
  def tengo_carta_libertad
    devolver = false
    if(@carta_libertad!=null)
      devolver = true
    end
    return devolver
  end
  
  def tengo_saldo(cantidad)
    devolver = false
    if(@saldo>=cantidad)
      devolver = true
    end
    return devolver
  end
  
  #def vender_propiedad(casilla)
  
  #end
  
  @Overrride
  def to_s()
    if(@propiedades.size==0)
      return "Nombre: #{@nombre}\n"\
              "Saldo: #{@saldo}\n"\
              "Encarcelado: #{@encarcelado}\n"\
              "Casilla actual: #{@casilla_actual}\n"\
              "Tiene carta libertad: #{@carta_libertad}\n"\
    else
      return "Nombre: #{@nombre}\n"\
           "Saldo: #{@saldo}\n"\
           "Capital: #{obtener_capital}\n"\
           "Encarcelado: #{@encarcelado}\n"\
           "Casilla actual: #{@casilla_actual}\n"\
           "Tiene carta libertad: #{@carta_libertad}\n"\
           "Propiedades: #{
              for i in 0.. @propiedades.size 
                @propiedades.to_s()
              end
          }\n"\
     end
  end
  
end
